Csv/misc bugfixes

tsmacdonald opened the pull request:
Fixes #30181 , which is just working through this list of problem CSVs: https://www.notion.so/metabase/9b4b30602feb40d7bb8405b1bdb76b75?v=a83d09fef7254d5b8ba796111c1f647d&p=c32a6331281d492fba06cef0539ef865&pm=s

There are three main problems:

* We were barfing on blank columns. This now replaces blank column names with `Unnamed Column X` (where `X` is the index of the column)
* Very long filenames were causing issues since the DB would silently truncate them and our `(t2/select Table :name the-name-we-provided)` was breaking. There were several ways I could've gone on this, but I opted to truncate the table name ourselves (so we can pretend we control it). There are of course different length limits for different DB engines :facepalm: . 
* Duplicate column names were causing problems. We now rename `foo, foo, foo` to `foo, foo-1, foo-2`.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/30219)
<!-- Reviewable:end -->

tsmacdonald committed:
Support empty header names

tsmacdonald committed:
Handle long CSV names more appropriately

Handle duplicate column names in CSVs

tsmacdonald committed:
Test fix

tsmacdonald committed:
Partition CSV inserts a little more intelligently

tsmacdonald committed:
Test fix

qnkhuat commented on src/metabase/driver/postgres.clj:
```diff
@@ -766,3 +766,8 @@
+
+(defmethod driver/table-name-length-limit :postgres
+  [_driver]
+  ;; https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS
```
per document this is not a constant but configurable via `NAMEDATALEN`.

should we take the extra miles and query it from DB instead?

otherwise, I think we should have a note in this implementation.

qnkhuat commented on src/metabase/driver/sql_jdbc.clj:
```diff
@@ -122,6 +122,13 @@
                                       :values      %})
-                        (partition-all 25 values))]
+                        (partition-all 100 values))]
+    ;; We need to partition the insert into multiple statements for both performance and correctness.
```
curious why correctness is concerned here?

tsmacdonald replied:
You can only insert so many values at a time. Postgres (and I think the other DBs are the same) only allows 65,535 parameterized values.

qnkhuat replied:
what does it do if we have more than that, does it throw an error or just siliently truncate it?

tsmacdonald replied:
By partitioning we (mostly) avoid the problem. If somebody has `> (65,534 / 100)` columns (so, 656 columns) they'll get a DB error. (Which is displayed nicely)

qnkhuat commented on src/metabase/upload.clj:
```diff
@@ -176,10 +193,12 @@
+  "Append the current datetime to the given name to create a unique table name. The resulting name will be short enough for the given driver (truncating the supplised `table-name` if necessary)."
+  [driver table-name]
+  (let [time-format "_yyyyMMddHHmmss"]
+    (str (subs (u/slugify table-name) 0 (min (count table-name)
```
there are just too many things we have to `deref` when reading this, can we break small things into meaningful `let` vars so it's easier to read?

things like `(t/format time-format (t/local-date-time))` could be called `date-time-postfix` would make it easier to read IMO.

qnkhuat commented on test/metabase/upload_test.clj:
```diff
@@ -296,12 +299,83 @@
                     alternating (map even? (range (count bool-column)))]
                 (is (= alternating bool-column))))))))))
 
+(deftest load-from-csv-length-test
```
I'd feel a lot safer if we test e2e this, the main reason is the `metabase.api.card/upload-csv!` is doing too many things and there are code paths we haven't tested there.

In this particular example, we are testing `unique-table-name` but from the call site, we do things a bit more complicated there.
https://github.com/metabase/metabase/blob/717a018dbc550032d55ed3fa968327efee511b7a/src/metabase/api/card.clj#L1007-L1008

calherries replied:
An e2e would ensure `unique-table-name` is being called on the table name. I think the confusion is this test just tests `unique-table-name`, but doesn't check that it's called. So you might as well change this to a unit test on `unique-table-name` without the `upload-load-from-csv` stuff.

calherries replied:
I decided to leave this how it is instead of changing it to a simple unit test on `unique-table-name` because this is also testing another property I didn't consider: that the database can handle the acceptable length of the column name given by `driver/table-name-length-limit`.

qnkhuat reviewed (changes requested):
Tests are failing and I also have some suggestions to improve readability

tsmacdonald committed:
Remove 'duplicate' from duplicated column names

tsmacdonald committed:
More intelligent sampling

tsmacdonald committed:
Add comment re. Postgres table-name-length-limit

tsmacdonald committed:
Quote column names, clean up tests

When unquoted, we couldn't use reserved words (true, false, group, etc.)
as column names

qnkhuat commented on resources/migrations/000_migrations.yaml:
```diff
@@ -14378,6 +14378,36 @@ databaseChangeLog:
+            tableName: metabase_table
+            columnName: display_name
+            newDataType: varchar(256)
+      rollback:
```
do we really need this rollback tho?

tsmacdonald replied:
The linter wanted it :shrug:

dpsutton replied:
I think you can leave an empty rollback if you like. ex,

```yaml
  - changeSet:
      id: v46.00-076
      author: noahmoss
      comment: Remove foreign key constraint on sandboxes.permission_id
      changes:
        - dropForeignKeyConstraint:
            baseTableName: sandboxes
            constraintName: fk_sandboxes_ref_permissions
      rollback:
      # Not required because a rollback would drop the `permission_id` column anyway
```

calherries replied:
True. By not having a rollback, the assumption we're making is that `varchar(256)` is fine for the last major version (which is almost certainly true).

tsmacdonald committed:
Fix test

Now that we quote things, H2 is case-sensitive

tsmacdonald committed:
Use dialect-appropriate quoting

Also clean up some card-filtering tests so they work with a dirty DB

calherries commented on src/metabase/upload.clj:
```diff
@@ -117,9 +117,26 @@
+      normalized
+      (format "unnamed_column_%s" (inc index)))))
+
+(defn- deduplicate
+  "Add `new-name` to the vector (must be a vector so that `conj` works!) of `names-so-far`, adding a unique suffix if
+  necessary."
+  [names-so-far new-name]
+  (if (some #(= % new-name) names-so-far)
+    (if-let [dupe-number (second (re-matches #".*_(\d+)" new-name))]
+      (recur names-so-far (str/replace new-name #"_\d+$" (str "_" (inc (parse-long dupe-number)))))
+      (recur names-so-far (str new-name "_1")))
+    (conj names-so-far new-name)))
```
Also, this will change the names of columns that are legitimately named `some_column_1`, which is not what we want. Either use `loop` and have an accumulator like `name-counts` keep track of how many occurrances of new-name there have been, or count how many dupes there are with `(count (filter #(= % new-name) names-so-far))`

tsmacdonald replied:
Recursion is necessary to handle admittedly-pathological cases:


```
(defn- cal-deduplicate
  "Add `new-name` to the vector (must be a vector so that `conj` works!) of `names-so-far`, adding a unique suffix if
  necessary."
  [names-so-far new-name]
  (let [new-name (if (some #(= % new-name) names-so-far)
                   (if-let [dupe-number (second (re-matches #".*_(\d+)" new-name))]
                     (str/replace new-name #"_\d+$" (str "_" (inc (parse-long dupe-number))))
                     (str new-name "_1"))
                   new-name)]
    (conj names-so-far new-name)))

(reduce cal-deduplicate [] ["a" "a" "a" "a_2"])
;; => ["a" "a_1" "a_1" "a_2"]

(reduce original-deduplicate [] ["a" "a" "a" "a_2"])
;; => ["a" "a_1" "a_2" "a_3"]
```

That said, while the `original-deduplicate` result above is _better_, it exemplifies your second point of clobbering columns that legitimately have a trailing number. I'm…working on it.

calherries replied:
True I understand now. If it sees an "a" and there is "a", "a_1", "a_2", etc you need to iterate until "a_N" is found. But yes, seems like you wouldn't need a recursive solution if you had an accumulator like `name-counts` though, as I think you figured out.

calherries commented on src/metabase/upload.clj:
```diff
@@ -117,9 +117,26 @@
+  (let [normalized (u/slugify (str/trim raw-name))]
+    (if-not (str/blank? normalized)
+      normalized
+      (format "unnamed_column_%s" (inc index)))))
```
`unnamed_column` should use the same deduplicate logic as normal columns. So you could just call it `unnamed_column` here and dedupe it later.

calherries commented on src/metabase/upload.clj:
```diff
@@ -176,11 +193,23 @@
+
+(def max-sample-rows "Maximum number of values to use for detecting a column's type" 1000)
+
+(defn- sample-rows
```
I'm guessing this is trying to solve performance problems with `rows->schema`. The problem is the data in the first N rows might not match the rest of the data. We should try to optimize `rows->schema` before resorting to this.

calherries replied:
I just read the [convo](https://metaboat.slack.com/archives/C04S696LRUM/p1681920630385549?thread_ts=1681908673.932979&cid=C04S696LRUM) on slack, never mind. Makes sense to do this vs the status quo

calherries commented on test/metabase/api/card_test.clj:
```diff
@@ -281,10 +281,11 @@
+            (let [recent-card-names (->> (mt/user-http-request :rasta :get 200 "card", :f :recent)
+                                         (map :name)
+                                         (filter #{"Card 1" "Card 2" "Card 3" "Card 4"}))]
+              (is (= ["Card 3" "Card 4" "Card 1"]
```
This is more concisely expressed with `(is (contains? ` and you could do without the `filter`.

tsmacdonald replied:
I had something like that at first, but I think preserving the order (3→4→1) is worthwhile.

calherries replied:
Ah, yes. Sorry. I remember you writing a test helper for this once. Something like this

```clojure
(defn contains-sequence [seq1 seq2]
  (loop [remaining-seq1 seq1
         remaining-seq2 seq2]
    (cond
      (empty? remaining-seq2) true
      (empty? remaining-seq1) false
      (= (first remaining-seq1) (first remaining-seq2))
      (recur (rest remaining-seq1) (rest remaining-seq2))
      :else (recur (rest remaining-seq1) seq2))))
```

Commit:
Fix deduplication

dpsutton commented on src/metabase/api/card.clj:
```diff
@@ -1001,14 +1001,14 @@ saved later when it is ready."
         driver            (driver.u/database->driver database)
         _                 (or (driver/database-supports? driver :uploads nil)
                               (throw (Exception. (tru "Uploads are not supported on {0} databases." (str/capitalize (name driver))))))
+        table-name        (->> (str (setting/get :uploads-table-prefix) filename-prefix)
```
why the use of `setting/get` instead of `(uploads-table-prefix)` as a function call?

calherries replied:
Good point, `setting/get` was leftover from something we used to have. Updated https://github.com/metabase/metabase/commit/ac66fddb91d190161695c1ddef76d56eb706cec7

Commit:
Remove use of setting/get

Commit:
Remove unneeded migration

tsmacdonald reviewed (commented):
I can't approve my own thing, but @calherries 's changes look good to me :heavy_check_mark: (thanks Cal!)
